addons/pkg/util/condition_util.go

```diff
@@ -0,0 +1,109 @@
+	stateExists bool
+	status      corev1.ConditionStatus
+}
+
+func (s *state) setState(cond v1alpha1.AppCondition) {
+	s.stateExists = true
+	if s.status == "" || cond.Status == corev1.ConditionTrue {
+		s.status = cond.Status
+	} else if s.status == corev1.ConditionUnknown && cond.Status == corev1.ConditionFalse {
+		s.status = cond.Status
+	}
+}
+
+// SummarizeAppConditions summarizes the provided conditions slice into a single condition with the following logic:
+// - If there be any 'Deleting' condition type, the summary condition type will be 'Deleting'
+// - Otherwise, if there be any 'Reconciling' condition type, the summary condition type will be 'Reconciling'
+// - Otherwise, if there be any 'DeleteFailed' condition type, the summary condition type will be 'DeleteFailed'
+// - Otherwise, if there be any 'ReconcileFailed' condition type, the summary condition type will be 'ReconcileFailed'
+// - Otherwise, if there be any 'ReconcileSucceeded' condition type, the summary condition type will be 'ReconcileSucceeded'
+// - Otherwise, the summary condition type will be 'UnknownCondition'
+// For the summary condition's Status field, it uses the logic in setState(), that is if there be any 'True' Status for a
+// given condition type, it overrides other existing 'False' or 'Unknown' Status entries for the same condition type and the
+// resulting condition will have 'True' Status. Similarly, existence of 'False' Status overrides 'Unknown' Status
+func SummarizeAppConditions(conditions []v1alpha1.AppCondition) v1alpha1.AppCondition {
+	summaryCondition := v1alpha1.AppCondition{}
+	var (
+		reconciling        = &state{}
+		reconcileFailed    = &state{}
+		reconcileSucceeded = &state{}
+		deleting           = &state{}
+		deleteFailed       = &state{}
+	)
+
+	for _, cond := range conditions {
+		if cond.Type == v1alpha1.Reconciling {
+			reconciling.setState(cond)
+		}
+		if cond.Type == v1alpha1.ReconcileFailed {
+			reconcileFailed.setState(cond)
+		}
+		if cond.Type == v1alpha1.ReconcileSucceeded {
+			reconcileSucceeded.setState(cond)
+		}
+		if cond.Type == v1alpha1.Deleting {
+			deleting.setState(cond)
+		}
+		if cond.Type == v1alpha1.DeleteFailed {
+			deleteFailed.setState(cond)
+		}
+	}
+
+	if deleting.stateExists {
+		summaryCondition.Type = v1alpha1.Deleting
+		summaryCondition.Status = deleting.status
+	} else if reconciling.stateExists {
+		summaryCondition.Type = v1alpha1.Reconciling
+		summaryCondition.Status = reconciling.status
+	} else if deleteFailed.stateExists {
+		summaryCondition.Type = v1alpha1.DeleteFailed
+		summaryCondition.Status = deleteFailed.status
+	} else if reconcileFailed.stateExists {
+		summaryCondition.Type = v1alpha1.ReconcileFailed
+		summaryCondition.Status = reconcileFailed.status
+	} else if reconcileSucceeded.stateExists {
+		summaryCondition.Type = v1alpha1.ReconcileSucceeded
+		summaryCondition.Status = reconcileSucceeded.status
+	} else {
+		summaryCondition.Type = UnknownCondition
+	}
```

Comment:
question: 
- Is it a valid case that `pkgi.status.conditions` has more than one item?
- If above is true, when the `pkgi.status.conditions` has ["ReconcileFailed", "ReoncileSucceeded"]. Will the summaryCondition is always set to `ReconcileFailed` ? (because the check for "ReconcileFailed" happens first)

Reply:
Yes for both questions

Reply:
>If above is true, when the pkgi.status.conditions has ["ReconcileFailed", "ReoncileSucceeded"]. Will the summaryCondition is always set to ReconcileFailed ? (because the check for "ReconcileFailed" happens first)

That makes me think if we want to summarize the condition as `ReoncileSuccess` ? (Based on the very last condition)

Reply:
I think ```"ReconcileFailed"``` and ```"ReoncileSucceeded"``` can not happen at the same time for a pkgi as they are both final states. But even if they did, the negative condition has more weight and we declare failure if we see it among the conditions

Reply:
I cannot recall if I have seen the following behavior from kapp-controller before:
- pkgi fails to be reconciled at the beginning due to `imagePullErr`
- kapp-controller retries it and pkgi succeeded the second time.
- We might want to check if ["reconcielFailed", "reconcileSucceeded"] could both exist under conditions.
  - if so, we probably want to populate only "reconcileSucceeded"
  - if not, will it be safe to assume the very last condition is the latest condition we would prefer to populate?

Reply:
What I learned from Carval Kapp controller's code, is that previous conditions are generally removed before adding a new one using ```removeAllCondition()``` but I definitely have seen multiple conditions at once. I checked with Dmitriy and he said we should not regard conditions slice as having history, so it's not that the most recent entry in the slice has the most recent or most correct information. @cppforlife could I ask you please elaborate on this and let me know if what I'm doing currently is right or needs change?

Reply:
Thanks Dmitriy. Changed logic @danniel1205